### PR TITLE
Add group end to end flow works.

### PR DESCRIPTION
**Pull request by Pushpalanka**

### CODE REVIEW

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties [hunk 1]
```diff
@@ -81,27 +82,34 @@ list.error.retrieve.profile=Failed to retrieve claim profile.
...
-profile.success=Successfully updated the profile!
-profile.failure=An Error occured when updating the profile!
-profile.updated=Your profile is updated.
-invalid.profile.name=Invalid profile name.
-profile.image.loaded=Profile image is loaded
-account.locked=Account is locked till user verify.
+groups.add.header=Add Group
+groups.edit.header=Edit Group
+group.edit.button=Update
+group.added.successfully.default=Group added successfully.
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties]: add group name to success message

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/add.js [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        sendToClient("action", null);
+        sendToClient("offset", null);
+        sendToClient("recordLimit", null);
+
+        var sortRowList = [];
+
+        return {
+            primaryDomainName: primaryDomainName,
+            domainNames: domainNames,
+            profile: getProfile(),
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/add.js]: need to handle error situation in getProfile()

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add-group.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        return;
+    }
+    var groupNameClaimUri = "http://wso2.org/claims/groupname";
+    var domain = null;
+    if (document.getElementById('domainSelector')) {
+        domain = document.getElementById('domainSelector').value;
+    }
+    $.ajax({
+        type: "GET",
+        url: "/admin-portal/root/apis/identityStore-micro-service/groupExists",
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add-group.js]: its better to remove hard corded root context admin-portal

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupBean.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+package org.wso2.is.portal.user.client.api.bean;
+
+/**
+ * Temporary GroupBean to be used in admin portal.
+ * <p>
+ * //todo enable cs
+ */
+public class GroupBean {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupBean.java]: add toString method to bean

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupListUserBean.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.is.portal.user.client.api.bean;
+
+/**
+ * Bean class used with group listing
+ */
+public class GroupListUserBean {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupListUserBean.java]: add toString method to bean

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 6]
```diff
@@ -299,6 +303,28 @@ public void updateUserProfile(String uniqueUserId, Map<String, String> updatedCl
...
+
+        List<Claim> updatedClaims = updatedClaimsMap.entrySet().stream()
+                .filter(entry -> !StringUtils.isNullOrEmpty(entry.getKey()))
+                .map(entry -> new Claim(IdentityMgtConstants.CLAIM_ROOT_DIALECT, entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+
+        try {
+            getRealmService().getIdentityStore().updateGroupClaims(uniqueGroupId, updatedClaims, null);
+        } catch (IdentityStoreException | GroupNotFoundException e) {
+            String error = "Failed to updated group profile.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: log context information

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 7]
```diff
@@ -320,6 +346,27 @@ public void updateUserProfile(String uniqueUserId, Map<String, String> updatedCl
...
+        List<Claim> claimList = null;
+
+        if (StringUtils.isNullOrEmpty(uniqueGroupId)) {
+            throw new UserPortalUIException("Invalid unique group id.");
+        }
+        if (metaClaims != null && !metaClaims.isEmpty()) {
+            try {
+                claimList = getRealmService().getIdentityStore().getClaimsOfGroup(uniqueGroupId, metaClaims);
+            } catch (IdentityStoreException | GroupNotFoundException e) {
+                String error = "Failed to get claims of the user.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: log context information

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 8]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+    }
+
+    @Override
+    public GroupBean addGroup(Map<String, String> groupClaims, String domainName) throws UserPortalUIException {
+
+        Group groupResult = null;
+        org.wso2.carbon.identity.mgt.bean.GroupBean groupBean = new org.wso2.carbon.identity.mgt.bean.GroupBean();
+        List<Claim> claimsList = new ArrayList<>();
+
+        for (Map.Entry<String, String> entry : groupClaims.entrySet()) {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: can use java8 lambda

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 9]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+            Claim claim = new Claim("http://wso2.org/claims", entry.getKey(), entry.getValue());
+            claimsList.add(claim);
+        }
+        groupBean.setClaims(claimsList);
+
+        try {
+            groupResult = getRealmService().getIdentityStore().addGroup(groupBean, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while adding groups.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context information to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 10]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+            throw new UserPortalUIException(error);
+        }
+        return new GroupBean(null, groupResult.getUniqueGroupId(), null, groupResult.getDomainName());
+    }
+
+    @Override
+    public boolean isGroupExist(Map<String, String> groupClaims, String domain) throws UserPortalUIException {
+        List<Claim> claimsList = new ArrayList<>();
+        boolean isGroupExists;
+        for (Map.Entry<String, String> entry : groupClaims.entrySet()) {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: can use java8 lambda

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 11]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+            Claim claim = new Claim();
+            claim.setClaimUri(entry.getKey());
+            claim.setValue(entry.getValue());
+            claimsList.add(claim);
+        }
+
+        try {
+            isGroupExists = getRealmService().getIdentityStore().isGroupExist(claimsList, domain);
+        } catch (IdentityStoreException e) {
+            String error = "Error while checking whether the group exists.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context informations to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 12]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+
+        List<Claim> updatedClaims = updatedClaimsMap.entrySet().stream()
+                .filter(entry -> !StringUtils.isNullOrEmpty(entry.getKey()))
+                .map(entry -> new Claim(IdentityMgtConstants.CLAIM_ROOT_DIALECT, entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+
+        try {
+            getRealmService().getIdentityStore().updateGroupClaims(uniqueGroupId, updatedClaims);
+        } catch (IdentityStoreException | GroupNotFoundException e) {
+            String error = "Failed to updated group profile.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context informations to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 13]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+    public List<UserListBean> getUserList(int offset, int length, String domainName,
+            List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<User> users;
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+        try {
+            users = getRealmService().getIdentityStore().listUsers(offset, length, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while retrieving users";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context information to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 14]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+    public List<GroupListUserBean> getUserListForAssignment(int offset, int length, String domainName,
+            List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<User> users;
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+        try {
+            users = getRealmService().getIdentityStore().listUsers(offset, length, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while retrieving users";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context information to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 15]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+
+        return generateUserListBeanForGroup(users, requestedClaims);
+    }
+
+    private List<GroupListUserBean> generateUserListBeanForGroup(List<User> users,
+            List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<GroupListUserBean> userList = new ArrayList<>();
+
+        List<MetaClaim> metaClaims = requestedClaims.stream()
+                .map(claimConfigEntry -> new MetaClaim("", claimConfigEntry.getClaimURI()))
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: why empty String ?

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 16]
```diff
@@ -501,4 +546,200 @@ private RealmService getRealmService() {
...
+            } catch (IdentityStoreException | UserNotFoundException e) {
+                String error = "Error while retrieving user data for user :  " + userEntry.getUniqueUserId();
+                LOGGER.error(error, e);
+                throw new UserPortalUIException(error);
+            }
+
+            GroupListUserBean listEntry = new GroupListUserBean();
+            listEntry.setPicture(null);           //set profile picture to be shown.
+            listEntry.setUid(userEntry.getUniqueUserId());
+            listEntry.setUsername(userClaims.get(0).getValue());
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Why 0 ?

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties (hunk 1 above)
Comment (Pushpalanka) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties]: Addressed.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/add.js (hunk 2 above)
Comment (Pushpalanka) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/add.js]: Addressed.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add-group.js (hunk 3 above)
Comment (Pushpalanka) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add-group.js]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 6 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed..

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 7 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 9 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 8 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 10 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 11 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 12 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 13 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 15 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed. Should have the default dialect.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java (hunk 16 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Addressed with iterating over the list and filtering the user claim. Zero was used as it was know there is only one claim and that is user claim as per the profile configuration. As this is not elegant corrected as pointed.

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupListUserBean.java (hunk 5 above)
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupListUserBean.java]: Addressed.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/index.hbs [hunk 17]
```diff
@@ -1,12 +1,50 @@
 {{secured}}
 {{layout "privilege"}}
 {{favicon "images/favicon.png" type="image/png"}}
-{{title "Users | " @config.appName}}
+{{title "Groups | " @config.appName}}
```
Comment (ruwanta) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/groups/index.hbs]: @GayanM  Why not i18n ?

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java [hunk 18]
```diff
@@ -147,16 +147,15 @@ public ProfileEntry getProfile(String profileName) throws UserPortalUIException
...
 
         List<Claim> claims;
         try {
             claims = getRealmService().getIdentityStore().getClaimsOfUser(uniqueUserId, metaClaims);
         } catch (IdentityStoreException e) {
             LOGGER.error(String.format("Failed to get the user claims for user - %s", uniqueUserId), e);
-            throw new UserPortalUIException(String.format("Failed to get the user claims for the profile - %s",
-                    profileName));
+            throw new UserPortalUIException(
+                    String.format("Failed to get the user claims for the profile - %s", profileName));
```
Comment (ruwanta) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java]: @GayanM Failed to get the user claims for the profile - %s, for user %s", profileName, uniqueUserId)

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java [hunk 19]
```diff
@@ -166,21 +165,63 @@ public ProfileEntry getProfile(String profileName) throws UserPortalUIException
...
+                .map(claimConfigEntry -> new MetaClaim(IdentityMgtConstants.CLAIM_ROOT_DIALECT,
+                        claimConfigEntry.getClaimURI())).collect(Collectors.toList());
+
+        List<Claim> claims;
+        try {
+            claims = getRealmService().getIdentityStore().getClaimsOfGroup(uniqueGroupId, metaClaims);
+        } catch (IdentityStoreException e) {
+            LOGGER.error(String.format("Failed to get the user claims for group - %s", uniqueGroupId), e);
+            throw new UserPortalUIException(
+                    String.format("Failed to get the group claims for the profile - group"));
```
Comment (ruwanta) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java]: @GayanM , 
1. Why log error here? 
2. Better to log the information about the meta-claims, if log.error is needed (1)
3. The "Failed to get the user claims for group" is misleading. What is the user here?

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java [hunk 20]
```diff
@@ -166,21 +165,63 @@ public ProfileEntry getProfile(String profileName) throws UserPortalUIException
...
+        } catch (GroupNotFoundException e) {
+            String error = String.format("Invalid group - %s", uniqueGroupId);
```
Comment (ruwanta) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/ProfileMgtClientServiceImpl.java]: String.format("The group - %s, not found while querying for the claims related to it", uniqueGroupId);

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupBean.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+
+    private static final long serialVersionUID = 56118812467L;
```
Comment (ruwanta) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/GroupBean.java]: @GayanM Why serialVersionUID ?

**Pull request merged**